{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Hannah Carter\n",
    "\n",
    "Purifying selection?\n",
    "Measures of linkage disequilibrium\n",
    "    D\n",
    "    R^2\n",
    "GWAS\n",
    "    Cases and controls you've identify\n",
    "    Want to identify loci that correlate with a phenotype in the population\n",
    "    There's some ancestral mutation that get inherited in the haplotype\n",
    "    Linked to certain haplotype\n",
    "    try to find loci that are correlated with the phenotype\n",
    "    don't need to know exactly what causes it\n",
    "    use snp chips\n",
    "        have like a million snps on the chip\n",
    "        use fluorescent incorporation to find particular SNP\n",
    "Estimating power to detect an association\n",
    "    Power = 1 -P(type II error)\n",
    "        type II error  = false negative\n",
    "    GPC power calculator\n",
    "    depends on\n",
    "        odds ratio\n",
    "        sample size\n",
    "        requred signficance\n",
    "        phenotype prevalence\n",
    "        extent to which tested marker is in LD with the causal marker\n",
    "Case-control\n",
    "    commonly used tests\n",
    "    Pearson's Chi^2\n",
    "        Calculate statistic from allelic dontingency table and assess significance\n",
    "        need at least 5 counts\n",
    "        basically estimate of fisher's exact test\n",
    "    Fisher's exact test\n",
    "        Assumed to follow a hypergeometric distribution\n",
    "        more computationally demanding\n",
    "        but can totally use it now\n",
    "    Logistic regression\n",
    "        Allows for inclusion of covariates\n",
    "Multiple genotypes are working together to give the phenotype\n",
    "QTL association testing\n",
    "    All regression based\n",
    "        Linear regression\n",
    "        most common model\n",
    "        easily include covariates\n",
    "        usually fit 1 marker, but can fit several at once\n",
    "        See if having a minor allele is asociated with a particular trait value\n",
    "    ANOVA\n",
    "        Similar to linear regression but focuses on comparing groups of samples\n",
    "        Analysis of variation between two or more samples\n",
    "        Lots of different types of models\n",
    "            fixed effects models\n",
    "            ranodm effects models\n",
    "            mixed effects models\n",
    "Evaluating Assocation Results\n",
    "    QQ-plots(quantile-quantile plots)\n",
    "        Plot observed statistic vs expected statistic both sorted in increasing order\n",
    "        Identify loci that deviate\n",
    "    Manhattan plot\n",
    "        plot -log10(pvalue) versus increasing genomic coordinate\n",
    "        usually use a p value of p<5e-8\n",
    "    LocusZoom\n",
    "        good tool for digging into what's going on"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Other considerations for GWAS\n",
    "Disease model\n",
    "    Additive\n",
    "        Assumes increase in effect from addition of each minor allele\n",
    "    Linear regresssion: 2 DF model\n",
    "Complicating factors\n",
    "    locus heterogeneity\n",
    "    pleiotropy\n",
    "    epistasis\n",
    "    environmental effects\n",
    "Environment\n",
    "    Phenotypic variance is caused by both genetic variance and environmental variance, but also by the interaction of the two\n",
    "Epistasis interaction between loci\n",
    "Tools for association testing\n",
    "    R, PLINK, GCTA\n",
    "    find stuff on bioconductor\n",
    "Population substructure\n",
    "    Effect sizes resulting from population level\n",
    "    Do things to control for it\n",
    "    Use PCA to look for population substructure\n",
    "Incomplete genealogies\n",
    "    based on assumptions of independent samples or samples with known relationships\n",
    "    if there some unaccounted for relatedness, it can invalidate your analyses\n",
    "    Need to measure this and remove highly related samples\n",
    "Mixed effects model\n",
    "    Instead of using principal components include a genetic relationship matrix as a random effect in a mixed effects model\n",
    "    strategy used by GC\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Quantitative Genetics\n",
    "    Classical model introduced by Ronald Fisher in 1918\n",
    "    You can estimate heritability\n",
    "Determining the role of common variation\n",
    "    GCTA\n",
    "    Mixed effects model incorporating genetic relationship matrix as a random effect\n",
    "    Fit phenotype to all SNPs at once\n",
    "    Estimate phenotypic variance explained by all SNPs\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
