{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Viewing alignments with IGV\n",
    "\n",
    "\n",
    "## Subset your bam file on only chr11\n",
    "\n",
    "To make the bam file smaller and easier to use, we'll use only the alignments from chromosome 11. \n",
    "\n",
    "### Exercise 1: use `samtools view` to subset your `S10` bam file on chromosome 11\n",
    "\n",
    "Notice that the usage output of `samtools view` shows you can specify a \"region\" such as \"`chr11`\" to get a subset of your data:\n",
    "\n",
    "```\n",
    "[ucsd-train01@tscc-login2 processed_data]$ samtools view -h\n",
    "\n",
    "Usage: samtools view [options] <in.bam>|<in.sam>|<in.cram> [region ...]\n",
    "```\n",
    "\n",
    "Go to your processed data folder and get only the \"`chr11`\" region. This will take a little bit.\n",
    "\n",
    "```\n",
    "[ucsd-train01@tscc-login2 ~]$ cd ~/projects/shalek2013/processed_data\n",
    "[ucsd-train01@tscc-login2 processed_data]$ samtools view [flag(s) to output as bam] S10.Aligned.out.sorted.bam [region] > S10.Aligned.out.sorted.chr11.bam\n",
    "```\n",
    "\n",
    "All programs using bam files also need the \"`.bai`\" index file, so we'll create that too.\n",
    "\n",
    "```\n",
    "[ucsd-train01@tscc-login2 processed_data]$ samtools index S10.Aligned.out.sorted.chr11.bam \n",
    "```\n",
    "\n",
    "You should now have these two files:\n",
    "\n",
    "```\n",
    "[ucsd-train01@tscc-login2 processed_data]$ ls -lha S10.Aligned.out.sorted.chr11.bam*\n",
    "-rw-r--r-- 1 ucsd-train01 biom262-group 418M Feb  4 06:20 S10.Aligned.out.sorted.chr11.bam\n",
    "-rw-r--r-- 1 ucsd-train01 biom262-group 119K Feb  4 06:23 S10.Aligned.out.sorted.chr11.bam.bai\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Secure copy the bam files to your laptop\n",
    "\n",
    "The file transfer takes about a minute.\n",
    "\n",
    "### Mac/Linux\n",
    "\n",
    "Use the `scp` secure copy command to copy files to your laptop. This is just like regular \"`cp oldfile newplace`\" except it knows how to deal with outside servers like TSCC. You have to specify the *exact* location of the file on TSCC.\n",
    "```\n",
    "scp ucsd-train##@tscc.sdsc.edu:~/projects/shalek2013/processed_data/S10.Aligned.out.sorted.chr11.bam ~/Desktop\n",
    "scp ucsd-train##@tscc.sdsc.edu:~/projects/shalek2013/processed_data/S10.Aligned.out.sorted.chr11.bam.bai ~/Desktop\n",
    "```\n",
    "\n",
    "This command will put the files on your desktop (since it says `~/Desktop` at the end).\n",
    "\n",
    "\n",
    "### Windows\n",
    "\n",
    "Add the directory with your putty executables to your path:\n",
    "\n",
    "```\n",
    "set PATH=C:\\path\\to\\putty\\directory;%PATH%\n",
    "```\n",
    "\n",
    "(you'll need to replace the fake path above with the real one on your computer - maybe it's `C:\\Olga\\Desktop`)\n",
    "\n",
    "Copy the files over, specifying your private key location, the location of the file on the server (specifying the exact location), and the place you want to put it on your computer.\n",
    "\n",
    "```\n",
    "pscp -i C:\\path\\to\\putty\\biom262\\private\\key ucsd-train##@tscc.sdsc.edu:~/projects/shalek2013/processed_data/S10.Aligned.out.sorted.chr11.bam %HOMEPATH%/Desktop\n",
    "pscp -i C:\\path\\to\\putty\\biom262\\private\\key ucsd-train##@tscc.sdsc.edu:~/projects/shalek2013/processed_data/S10.Aligned.out.sorted.chr11.bam.bai %HOMEPATH%/Desktop\n",
    "```\n",
    "\n",
    "This command will put the files on your desktop (since it says `%HOMEPATH%/Desktop` at the end). Let me know if this doesn't work."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the Integrated Genome Viewer (IGV) to view alignments\n",
    "\n",
    "### Open IGV\n",
    "\n",
    "Find where you downloaded [IGV](https://www.broadinstitute.org/software/igv/log-in) (click there to download if you haven't yet. it'll ask you to login/register because they need to track users for their grants), and double-click it to open.\n",
    "\n",
    "### Choose the correct genome\n",
    "\n",
    "These data were aligned to the `mm10` version of the genome, but the default on IGV is `hg19`. So we'll need to change the genome. In the upper left where it says \"Human hg19\", click that and then click \"More ...\"\n",
    "\n",
    "![Choose genome from IGV](figures/igv_choose_genome.png)\n",
    "\n",
    "In the genome box, start typing \"mm10\" to search and get the genome.\n",
    "\n",
    "**You won't need to search for the genome every time -- just this first time.** But you'll always need to make sure your genome is correct :)\n",
    "\n",
    "![Search genome list](figures/igv_search_genome_list.png)\n",
    "\n",
    "### Load your bam file\n",
    "\n",
    "Choose \"File > Load from file.\"\n",
    "\n",
    "![Load file](figures/igv_load_from_file.png)\n",
    "\n",
    "Find your the bam file that you just downloaded. Once loaded, your screen should look like:\n",
    "\n",
    "![Loaded data](figures/igv_loaded_data.png)\n",
    "\n",
    "### Go to your gene of interest\n",
    "\n",
    "Two of the genes from the paper, *Irgm1* and *Irf7* are on chromosome 11. Let's check them out. Search for \"Irgm1\" in the search box and press \"Enter.\"\n",
    "\n",
    "![Search Irgm1](figures/igv_search_irgm1_gene.png)\n",
    "\n",
    "Now you should see the *Irgm1* locus:\n",
    "\n",
    "![Irgm1 locus](figures/igv_irgm1_locus.png)\n",
    "\n",
    "\n",
    "### Show the splice junction track\n",
    "\n",
    "#### Hover over the splice junction track to see the number of junction reads (\"Depth\")\n",
    "\n",
    "### Zoom in to the 5' UTR\n",
    "\n",
    "#### Look at the read properties\n",
    "\n",
    "#### See SNPs/errors in the reads"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
